sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompositeOffset.scala

```diff
@@ -52,6 +52,21 @@ case class CompositeOffset(offsets: Seq[Option[Offset]]) extends Offset {
     case i if i == 0 => 0
     case i if i > 0 => 1
   }
+
+  /**
+   * Unpacks an offset into [[StreamProgress]] by associating each offset with the order list of
+   * sources.
+   *
+   * This method is typically used to associate a serialized offset with actual sources (which
+   * cannot be serialized).
+   */
+  def toStreamProgress(
+      sources: Seq[Source],
+      dest: StreamProgress = new StreamProgress): StreamProgress = {
```

Comment:
I made this all immutable (so it is actually constructing a new one always).  I also got rid of the now unneeded locking.